### PR TITLE
CompositeDistribution: Fix getSingularities and add computeProbability

### LINKED ISSUE #3308: CompositeDistribution: getSingularities and computeProbability bugs
## Description

`CompositeDistribution` (push-forward of a 1D distribution through a scalar function `f: R→R`) has several bugs in `getSingularities()` and `computeProbability()` that produce incorrect results for distributions with interior critical points or singular PDF endpoints.

These bugs are blocking for any downstream use that relies on numerical integration over the support (entropy, roughness, probability intervals, level sets).

## Bug 1 — `getSingularities()` returns wrong values (major)

**File:** `CompositeDistribution.cxx:467-477`

The current implementation sorts all `values_` (the images of the partition bounds) and removes the first and last elements, assuming they are the range endpoints `f(xMin)` and `f(xMax)`. This is wrong when an **interior extremum** achieves the global min or max.

### Examples

| f(x) | Antecedent | Returned | Expected |
|---|---|---|---|
| x² | N(0,1) | [58.53] (=f(xMax)) | [0] (=f(0), the interior minimum) |
| sin | U(0,2π) | [±1e-12] (~f(xMin)≈f(xMax)≈0) | [-1, +1] (interior extrema) |
| exp(x) | N(0,1) | [0] (spurious) | [] (empty, no interior critical point) |
| x (identity) | N(0,1) | [0] (spurious) | [] (empty) |

**Root cause:** `erase(0)` and `erase(last)` after sorting removes the *smallest and largest values* in the array, not the *endpoint images* `f(xMin)` and `f(xMax)`. When the global extremum is achieved at an interior critical point (e.g., 0 for x², ±1 for sin), the true singular point is deleted and a spurious one is kept.

**Fix:** Return `values_[1..size-2]` (the images of the interior partition bounds = the critical values of f), sorted ascending for use by the base class integration routines. For the monotone case (`values_.size() == 2`), return an empty `Point()`.

### Consequences

- `computeEntropy()` returns **inf** for `exp(N(0,1))` because the spurious singularity at 0 (below range lower) creates a degenerate integration sub-interval.
- `computeRoughness()` (via `getRoughness()`) and `computeProbabilityContinuous1D()` split the integration interval at the wrong points, leading to inaccurate results.
- GaussKronrod integration over the full range of `sin(U(0,2π))` returns **117.38** instead of 1.0 (the singular endpoints ±1 are not identified as split points).

## Bug 2 — `computeProbability()` not overridden (major)

**File:** `CompositeDistribution.cxx` (missing override)

`CompositeDistribution` does not override `computeProbability(const Interval &)`. The base class `DistributionImplementation::computeProbabilityContinuous1D()` integrates the PDF via GaussKronrod, which **diverges** at range-endpoint singularities.

### Examples (sin(X), X~U(0,2π), range=[-1, 1])

| Interval | Returned | Expected (MC) |
|---|---|---|
| [-0.99, -0.5] | 0.2883 | 0.2887 ✓ |
| [-1.0, 0.0] | **1.0** | 0.4994 ✗ |
| [-1.0, -0.5] | **1.0** | 0.2113 ✗ |
| [0.5, 1.0] | **1.0** | 0.3325 ✗ |
| [-1.0, 1.0] | 1.0 | 1.0 ✓ (accident) |

The intervals that touch a range endpoint where the PDF is singular return 1.0 regardless of the actual probability.

**Fix:** Override `computeProbability` using a segment-wise summation that mirrors the logic of `computeCDF` (reusing `bounds_`, `values_`, `probabilities_`, `increasing_`), but adapted for an arbitrary interval `[a, b]` rather than `(-∞, x]`. This avoids catastrophic cancellation from `CDF(b) - CDF(a)` when both are close to 1.

## Bug 3 — Spurious singularity for monotone functions (minor)

`getSingularities()` returns `Point(1, 0)` when `values_.getSize() == 2` (monotone case). This claims y=0 is a singularity when the PDF is smooth everywhere. Should return an empty `Point()`.

**Consequence:** Unnecessary split at y=0 in integration routines. Benign for smooth PDFs but semantically incorrect and could interact badly with `AdaptiveStieltjesAlgorithm`.

## Bug 4 — Entropy inf/nan (consequence of Bug 1)

`computeEntropy()` of `exp(N(0,1))` returns `inf` because the spurious singularity at 0 (below range lower bound ~0.000475) causes a degenerate sub-interval in the GK integration.

The correct entropy of `exp(N(0,1))` is ≈1.4189 (log-normal with μ=0, σ=1).

## Environment

- OpenTURNS 1.28dev (manifold_distribution branch, commit 817c1f9a6)
- Reproduced on Linux x86_64, Python 3.13
- `CompositeDistribution-StepNumber` = 256, `CompositeDistribution-SolverEpsilon` = 1e-14

## Suggested fix summary

1. **`getSingularities()`**: return `values_[1..size-2]` (sorted), or empty Point if size ≤ 2.
2. **`computeProbability(Interval)`**: new override using segment-wise summation over `bounds_`/`values_`/`probabilities_`/`increasing_`, analogous to `computeCDF` but for an arbitrary interval `[a, b]`.
3. **Remove spurious `Point(1, 0)`** for the monotone case.

### PR DESCRIPTION
**Pull request by regislebrun**
## Description

`CompositeDistribution` (push-forward of a 1D distribution through a scalar function `f: R→R`) has several bugs in `getSingularities()` and `computeProbability()` that produce incorrect results for distributions with interior critical points or singular PDF endpoints.

These bugs are blocking for any downstream use that relies on numerical integration over the support (entropy, roughness, probability intervals, level sets).

## Bug 1 — `getSingularities()` returns wrong values (major)

**File:** `CompositeDistribution.cxx:467-477`

The current implementation sorts all `values_` (the images of the partition bounds) and removes the first and last elements, assuming they are the range endpoints `f(xMin)` and `f(xMax)`. This is wrong when an **interior extremum** achieves the global min or max.

### Examples

| f(x) | Antecedent | Returned | Expected |
|---|---|---|---|
| x² | N(0,1) | [58.53] (=f(xMax)) | [0] (=f(0), the interior minimum) |
| sin | U(0,2π) | [±1e-12] (~f(xMin)≈f(xMax)≈0) | [-1, +1] (interior extrema) |
| exp(x) | N(0,1) | [0] (spurious) | [] (empty, no interior critical point) |
| x (identity) | N(0,1) | [0] (spurious) | [] (empty) |

**Root cause:** `erase(0)` and `erase(last)` after sorting removes the *smallest and largest values* in the array, not the *endpoint images* `f(xMin)` and `f(xMax)`. When the global extremum is achieved at an interior critical point (e.g., 0 for x², ±1 for sin), the true singular point is deleted and a spurious one is kept.

**Fix:** Return `values_[1..size-2]` (the images of the interior partition bounds = the critical values of f), sorted ascending for use by the base class integration routines. For the monotone case (`values_.size() == 2`), return an empty `Point()`.

### Consequences

- `computeEntropy()` returns **inf** for `exp(N(0,1))` because the spurious singularity at 0 (below range lower) creates a degenerate integration sub-interval.
- `computeRoughness()` (via `getRoughness()`) and `computeProbabilityContinuous1D()` split the integration interval at the wrong points, leading to inaccurate results.
- GaussKronrod integration over the full range of `sin(U(0,2π))` returns **117.38** instead of 1.0 (the singular endpoints ±1 are not identified as split points).

## Bug 2 — `computeProbability()` not overridden (major)

**File:** `CompositeDistribution.cxx` (missing override)

`CompositeDistribution` does not override `computeProbability(const Interval &)`. The base class `DistributionImplementation::computeProbabilityContinuous1D()` integrates the PDF via GaussKronrod, which **diverges** at range-endpoint singularities.

### Examples (sin(X), X~U(0,2π), range=[-1, 1])

| Interval | Returned | Expected (MC) |
|---|---|---|
| [-0.99, -0.5] | 0.2883 | 0.2887 ✓ |
| [-1.0, 0.0] | **1.0** | 0.4994 ✗ |
| [-1.0, -0.5] | **1.0** | 0.2113 ✗ |
| [0.5, 1.0] | **1.0** | 0.3325 ✗ |
| [-1.0, 1.0] | 1.0 | 1.0 ✓ (accident) |

The intervals that touch a range endpoint where the PDF is singular return 1.0 regardless of the actual probability.

**Fix:** Override `computeProbability` using a segment-wise summation that mirrors the logic of `computeCDF` (reusing `bounds_`, `values_`, `probabilities_`, `increasing_`), but adapted for an arbitrary interval `[a, b]` rather than `(-∞, x]`. This avoids catastrophic cancellation from `CDF(b) - CDF(a)` when both are close to 1.

## Bug 3 — Spurious singularity for monotone functions (minor)

`getSingularities()` returns `Point(1, 0)` when `values_.getSize() == 2` (monotone case). This claims y=0 is a singularity when the PDF is smooth everywhere. Should return an empty `Point()`.

**Consequence:** Unnecessary split at y=0 in integration routines. Benign for smooth PDFs but semantically incorrect and could interact badly with `AdaptiveStieltjesAlgorithm`.

## Bug 4 — Entropy inf/nan (consequence of Bug 1)

`computeEntropy()` of `exp(N(0,1))` returns `inf` because the spurious singularity at 0 (below range lower bound ~0.000475) causes a degenerate sub-interval in the GK integration.

The correct entropy of `exp(N(0,1))` is ≈1.4189 (log-normal with μ=0, σ=1).

## Environment

- OpenTURNS 1.28dev (manifold_distribution branch, commit 817c1f9a6)
- Reproduced on Linux x86_64, Python 3.13
- `CompositeDistribution-StepNumber` = 256, `CompositeDistribution-SolverEpsilon` = 1e-14

## Suggested fix summary

1. **`getSingularities()`**: return `values_[1..size-2]` (sorted), or empty Point if size ≤ 2.
2. **`computeProbability(Interval)`**: new override using segment-wise summation over `bounds_`/`values_`/`probabilities_`/`increasing_`, analogous to `computeCDF` but for an arbitrary interval `[a, b]`.
3. **Remove spurious `Point(1, 0)`** for the monotone case.

Closes #3308

## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected singularity detection for composite distributions.
  - Improved interval probability calculations, including cases near PDF singularities and outside the distribution range.
  - Prevented potential divergence during interval probability computation.

- **Documentation**
  - Added guidance on composite distribution ranges, singularities, and interval probability calculations.

- **Tests**
  - Expanded coverage for probability calculations, singularities, serialization, edge cases, and distribution behavior.

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -41,6 +41,9 @@
  * Add Process sample slicing operator
  * Re-raise native Python exceptions through PythonFunction instead of RuntimeException
 
+=== Bug fixes (total 1) ===
+ * #3308 (The CompositeDistribution.getSingularities method returns incorrect values and the computation of the interval probabilities can diverge)
```
Comment (jschueller) [inline ChangeLog]: drop it, we'll update this on release

File: python/test/t_CompositeDistribution_std.py [hunk 2]
```diff
@@ -167,3 +168,198 @@
...
+d_assign = ot.CompositeDistribution(d_copy)
+d_assign.setFunction(f2)
+assert d_assign != d1, "setFunction should update the distribution"
+
+# Constructor argument checks
+with ott.assert_raises(Exception):
+    ot.CompositeDistribution(ot.SymbolicFunction(["x", "y"], ["x+y"]), ot.Normal())  # wrong input dim
+with ott.assert_raises(Exception):
+    ot.CompositeDistribution(ot.SymbolicFunction("x", ["x", "x^2"]), ot.Normal())  # wrong output dim
+with ott.assert_raises(Exception):
```
Comment (jschueller) [inline python/test/t_CompositeDistribution_std.py]: TypeError ? RuntimeError ? Exception is too generic

**Review by jschueller — APPROVED** [review]

File: python/test/t_CompositeDistribution_std.py [hunk 3]
```diff
@@ -167,3 +168,198 @@
...
+# x^3-3x: f'(x)=3x^2-3=0 at x=+/-1, singularities at y=f(-1)=2, y=f(1)=-2
+f_pl = ot.SymbolicFunction("x", "x^3 - 3*x")
+d_pl = ot.CompositeDistribution(f_pl, ot.Normal())
+sing_pl = d_pl.getSingularities()
+ott.assert_almost_equal(sing_pl.getSize(), 2)
+ott.assert_almost_equal(sing_pl, [-2.0, 2.0])  # sorted ascending
+
+# sin: f'(x)=cos(x)=0 at pi/2, 3pi/2, singularities at y=...=+/-1
+f_sin = ot.SymbolicFunction("x", "sin(x)")
+d_sin = ot.CompositeDistribution(f_sin, ot.Uniform(0, 6.283185307))
```
Comment (jschueller) [inline python/test/t_CompositeDistribution_std.py]: use 2 * math.pi

File: python/test/t_CompositeDistribution_std.py [hunk 4]
```diff
@@ -167,3 +168,198 @@
...
+d_mut.setAntecedent(ot.Normal(2.0, 1.0))
+ott.assert_almost_equal(d_mut.getMean()[0], chi2.getMean()[0] + 4.0, 1e-2, 1e-2)
+d_mut.setFunction(f_id)
+ott.assert_almost_equal(d_mut.getMean()[0], 2.0, 1e-3, 1e-3)
+# setFunction to a function with no critical points removes singularities
+d_mut.setFunction(f_exp)
+ott.assert_almost_equal(d_mut.getSingularities().getSize(), 0)
+
+# setSolver changes the solver
+d_solv = ot.CompositeDistribution(f_sin, ot.Uniform(0.0, 6.283185307))
```
Comment (jschueller) [inline python/test/t_CompositeDistribution_std.py]: here too

File: python/test/t_CompositeDistribution_std.py [hunk 5]
```diff
@@ -167,3 +168,198 @@
...
+
+# sin P[-0.99,-0.5] should be finite and strictly below 1 (the old GK bug gave 1.0)
+p_sin3 = d_sin.computeProbability(ot.Interval([-0.99], [-0.5]))
+assert p_sin3 > 0.0 and p_sin3 < 1.0, f"sin P[-0.99,-0.5]={p_sin3} should be in (0,1)"
+
+# computePDF must sum to 1 when integrated over its full range:
+# PDF at 0.5 for sin: 2 preimages, each with |cos(x)|=sqrt(3)/2, pdf=1/(2pi)*(2/|cos|)=4/(2pi sqrt(3))
+ott.assert_almost_equal(
+    d_sin.computePDF(0.5),
+    4.0 / (2.0 * 6.283185307 * 0.8660254037844387),
```
Comment (jschueller) [inline python/test/t_CompositeDistribution_std.py]: use exact value

File: python/test/t_CompositeDistribution_std.py [hunk 6]
```diff
@@ -1,9 +1,12 @@
 #! /usr/bin/env python
 
+import pickle
+
 import openturns as ot
 import openturns.testing as ott
 
 ot.TESTPREAMBLE()
+ot.RandomGenerator.SetSeed(0)
```
Comment (jschueller) [inline python/test/t_CompositeDistribution_std.py]: can be dropped I think

File: ChangeLog (hunk 1 above)
Comment (regislebrun) [reply ChangeLog]: Done. Removed the Bug fixes section - it will be updated at release time.

File: python/test/t_CompositeDistribution_std.py (hunk 2 above)
Comment (regislebrun) [reply python/test/t_CompositeDistribution_std.py]: Done. Used TypeError, which matches the Python exception type surfaced by the SWIG wrapper for all three constructor argument checks.

File: lib/src/Uncertainty/Distribution/CompositeDistribution.cxx [hunk 7]
```diff
@@ -440,6 +440,83 @@ Scalar CompositeDistribution::computeCDF(const Point & point) const
...
 }
 
+/* Get the probability content of an interval */
+Scalar CompositeDistribution::computeProbability(const Interval & interval) const
+{
+  if (interval.getDimension() != 1) throw InvalidArgumentException(HERE) << "Error: the given interval must have dimension 1, here dimension=" << interval.getDimension();
+  Scalar probability = 0.0;
+  const Scalar lo = interval.getLowerBound()[0];
+  const Scalar hi = interval.getUpperBound()[0];
+  if (!(hi > lo)) return probability;
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/CompositeDistribution.cxx]: Done. Added handling for constant segments (fA == fB): the atom mass is now counted when the interval contains the constant value.

File: lib/src/Uncertainty/Distribution/CompositeDistribution.cxx [hunk 8]
```diff
@@ -466,13 +543,15 @@ LevelSet CompositeDistribution::computeMinimumVolumeLevelSetWithThreshold(const
...
-  // The singularities are at the extrema of f
-  Point singularities(values_);
+  // The singularities are at the images of the roots of the derivative
+  // of f located inside of the range, ie the interior elements of values_.
+  // The first and last elements of values_ are the images of the bounds
+  // of the range, which are not singularities.
+  const UnsignedInteger size = values_.getSize();
+  if (size <= 2) return Point(0);
+  Point singularities(size - 2);
+  for (UnsignedInteger i = 1; i < size - 1; ++i) singularities[i - 1] = values_[i];
```
[the remark this answers is not in the document]
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/CompositeDistribution.cxx]: Done. getSingularities now only returns images of interior points where the monotonicity flag actually changes, so refinement points inside a monotonic region are excluded.

File: python/test/t_CompositeDistribution_std.py (hunk 3 above)
Comment (regislebrun) [reply python/test/t_CompositeDistribution_std.py]: Done.

File: python/test/t_CompositeDistribution_std.py (hunk 4 above)
Comment (regislebrun) [reply python/test/t_CompositeDistribution_std.py]: Done.

File: python/test/t_CompositeDistribution_std.py (hunk 5 above)
Comment (regislebrun) [reply python/test/t_CompositeDistribution_std.py]: Done. Replaced the numeric constants with exact algebraic expressions.

File: python/test/t_CompositeDistribution_std.py (hunk 6 above)
Comment (regislebrun) [reply python/test/t_CompositeDistribution_std.py]: Done. Dropped SetSeed(0) - the default RNG seed is already 0 via RandomGenerator-InitialSeed.
